File: DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/itemencryptor/internaldafny/legacy/InternalLegacyOverride.java
package software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.internaldafny.legacy;

/**
 * This file does *NOT* import a lot of things.
 * This is because it is dealing with converting
 * between different versions of the same name.
 * The DynamoDbItemEncryptor module has Dafny and Java versions
 * of the same type.
 * This means that `EncryptItemOutput` for example
 * needs to be disambiguated between the Dafny version and the Java version.
 * In order to make it clearer at each call-site exactly what is happening
 * the full import is used.
 * IDEs tend to fight this so I'm sorry.
 */

import StandardLibraryInternal.InternalResult;
import Wrappers_Compile.Option;
import Wrappers_Compile.Result;
import com.amazonaws.services.dynamodbv2.datamodeling.sdkv2.encryption.EncryptionContext;
import com.amazonaws.services.dynamodbv2.datamodeling.sdkv2.encryption.EncryptionFlags;
import dafny.DafnyMap;
import dafny.DafnySequence;
import dafny.TypeDescriptor;
import java.util.*;
import java.util.function.BiConsumer;
import software.amazon.cryptography.dbencryptionsdk.dynamodb.ILegacyDynamoDbEncryptor;
import software.amazon.cryptography.dbencryptionsdk.dynamodb.internaldafny.types.LegacyPolicy;
import software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.internaldafny.types.Error;
import software.amazon.cryptography.dbencryptionsdk.structuredencryption.internaldafny.types.CryptoAction;

public class InternalLegacyOverride extends _ExternBase_InternalLegacyOverride {

  private final LegacyEncryptorAdapter _encryptorAdapter;
  private final LegacyPolicy _policy;
  private final DafnySequence<Character> materialDescriptionFieldNameDafny;
  private final DafnySequence<Character> signatureFieldNameDafny;

  private InternalLegacyOverride(
    LegacyEncryptorAdapter encryptorAdapter,
    LegacyPolicy policy
  ) {
    this._encryptorAdapter = encryptorAdapter;
    this._policy = policy;
    // It is possible that these values
    // have been customized by the customer.
    this.materialDescriptionFieldNameDafny =
      software.amazon.smithy.dafny.conversion.ToDafny.Simple.CharacterSequence(
        encryptorAdapter.getMaterialDescriptionFieldName()
      );
    this.signatureFieldNameDafny =
      software.amazon.smithy.dafny.conversion.ToDafny.Simple.CharacterSequence(
        encryptorAdapter.getSignatureFieldName()
      );
  }

  public static TypeDescriptor<InternalLegacyOverride> _typeDescriptor() {
    return TypeDescriptor.referenceWithInitializer(
      InternalLegacyOverride.class,
      () -> null
    );
  }

  public boolean IsLegacyInput(
    software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.internaldafny.types.DecryptItemInput input
  ) {
    //= specification/dynamodb-encryption-client/decrypt-item.md#determining-legacy-items
    //# An item MUST be determined to be encrypted under the legacy format if it contains
    //# attributes for the material description and the signature.
    return (
      input.is_DecryptItemInput() &&
      input._encryptedItem.contains(materialDescriptionFieldNameDafny) &&
      input._encryptedItem.contains(signatureFieldNameDafny)
    );
  }

  public LegacyPolicy policy() {
    return _policy;
  }

  public Result<
    software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.internaldafny.types.EncryptItemOutput,
    Error
  > EncryptItem(
    software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.internaldafny.types.EncryptItemInput input
  ) {
    // Precondition: Policy MUST allow the caller to encrypt.
    if (!_policy.is_FORCE__LEGACY__ENCRYPT__ALLOW__LEGACY__DECRYPT()) {
      return CreateEncryptItemFailure(
        createError("Legacy Policy does not support encrypt.")
      );
    }

    try {
      Map<
        String,
        software.amazon.awssdk.services.dynamodb.model.AttributeValue
      > plaintextItem =
        software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.ToNative
          .EncryptItemInput(input)
          .plaintextItem();

      final Map<
        String,
        software.amazon.awssdk.services.dynamodb.model.AttributeValue
      > encryptedItem = _encryptorAdapter.encryptRecord(plaintextItem);

      final software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.model.EncryptItemOutput nativeOutput =
        software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.model.EncryptItemOutput
          .builder()
          .encryptedItem(encryptedItem)
          .build();
      final software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.internaldafny.types.EncryptItemOutput dafnyOutput =
        software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.ToDafny.EncryptItemOutput(
          nativeOutput
        );
      return CreateEncryptItemSuccess(dafnyOutput);
    } catch (Exception ex) {
      return CreateEncryptItemFailure(Error.create_Opaque(ex));
    }
  }

  public Result<
    software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.internaldafny.types.DecryptItemOutput,
    Error
  > DecryptItem(
    software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.internaldafny.types.DecryptItemInput input
  ) {
    // Precondition: Policy MUST allow the caller to decrypt.
    //= specification/dynamodb-encryption-client/decrypt-item.md#behavior
    //# If a [Legacy Policy](./ddb-table-encryption-config.md#legacy-policy) of
    //# `FORBID_LEGACY_ENCRYPT_FORBID_LEGACY_DECRYPT` is configured,
    //# and the input item [is an item written in the legacy format](#determining-legacy-items),
    //# this operation MUST fail.
    if (
      !_policy.is_FORCE__LEGACY__ENCRYPT__ALLOW__LEGACY__DECRYPT() &&
      !_policy.is_FORBID__LEGACY__ENCRYPT__ALLOW__LEGACY__DECRYPT()
    ) {
      return CreateDecryptItemFailure(
        createError("Legacy Policy does not support decrypt.")
      );
    }
    try {
      Map<
        String,
        software.amazon.awssdk.services.dynamodb.model.AttributeValue
      > encryptedItem =
        software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.ToNative
          .DecryptItemInput(input)
          .encryptedItem();

      Map<
        String,
        software.amazon.awssdk.services.dynamodb.model.AttributeValue
      > plaintextItem = _encryptorAdapter.decryptRecord(encryptedItem);

      final software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.model.DecryptItemOutput nativeOutput =
        software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.model.DecryptItemOutput
          .builder()
          .plaintextItem(plaintextItem)
          .build();
      final software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.internaldafny.types.DecryptItemOutput dafnyOutput =
        software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.ToDafny.DecryptItemOutput(
          nativeOutput
        );
      return CreateDecryptItemSuccess(dafnyOutput);
    } catch (Exception ex) {
      return CreateDecryptItemFailure(Error.create_Opaque(ex));
    }
  }

  public static Result<Option<InternalLegacyOverride>, Error> Build(
    software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.internaldafny.types.DynamoDbItemEncryptorConfig encryptorConfig
  ) {
    // Check for early return (Postcondition): If there is no legacyOverride there is nothing to do.
    if (encryptorConfig.dtor_legacyOverride().is_None()) {
      return CreateBuildSuccess(CreateInternalLegacyOverrideNone());
    }
    final software.amazon.cryptography.dbencryptionsdk.dynamodb.internaldafny.types.LegacyOverride legacyOverride =
      encryptorConfig.dtor_legacyOverride().dtor_value();

    final ILegacyDynamoDbEncryptor maybeEncryptor =
      software.amazon.cryptography.dbencryptionsdk.dynamodb.ToNative.LegacyDynamoDbEncryptor(
        legacyOverride.dtor_encryptor()
      );

    // Precondition: The encryptor MUST be a DynamoDBEncryptor
    if (!isDynamoDBEncryptor(maybeEncryptor)) {
      return CreateBuildFailure(
        createError("Legacy encryptor is not supported")
      );
    }
    // Preconditions: MUST be able to create valid encryption context
    final InternalResult<EncryptionContext, Error> maybeEncryptionContext =
      legacyEncryptionContext(encryptorConfig);
    if (maybeEncryptionContext.isFailure()) {
      return CreateBuildFailure(maybeEncryptionContext.error());
    }
    // Precondition: All actions MUST be supported types
    final InternalResult<
      Map<String, Set<EncryptionFlags>>,
      Error
    > maybeActions = legacyActions(
      legacyOverride.dtor_attributeActionsOnEncrypt()
    );
    if (maybeActions.isFailure()) {
      return CreateBuildFailure(maybeEncryptionContext.error());
    }

    final LegacyEncryptorAdapter encryptorAdapter;
    if (
      maybeEncryptor instanceof
      com.amazonaws.services.dynamodbv2.datamodeling.sdkv2.encryption.DynamoDBEncryptor
    ) {
      encryptorAdapter =
        new V2EncryptorAdapter(
          (com.amazonaws.services.dynamodbv2.datamodeling.sdkv2.encryption.DynamoDBEncryptor) maybeEncryptor,
          maybeActions.value(),
          maybeEncryptionContext.value()
        );
    } else {
      return CreateBuildFailure(
        createError(
          "Unsupported encryptor type: " + maybeEncryptor.getClass().getName()
        )
      );
    }

    final InternalLegacyOverride internalLegacyOverride =
      new InternalLegacyOverride(
        encryptorAdapter,
        legacyOverride.dtor_policy()
      );

    return CreateBuildSuccess(
      CreateInternalLegacyOverrideSome(internalLegacyOverride)
    );
  }

  //  Everything below this point is an implementation detail

  public static Error createError(String message) {
    final DafnySequence<Character> dafnyMessage =
      software.amazon.smithy.dafny.conversion.ToDafny.Simple.CharacterSequence(
        message
      );
    return Error.create_DynamoDbItemEncryptorException(dafnyMessage);
  }

  public static boolean isDynamoDBEncryptor(
    software.amazon.cryptography.dbencryptionsdk.dynamodb.ILegacyDynamoDbEncryptor maybe
  ) {
    return (
      maybe instanceof
        com.amazonaws.services.dynamodbv2.datamodeling.sdkv2.encryption.DynamoDBEncryptor
    );
  }

  public static String ToNativeString(DafnySequence<? extends Character> s) {
    return software.amazon.smithy.dafny.conversion.ToNative.Simple.String(s);
  }

  public static DafnySequence<Character> ToDafnyString(String s) {
    return software.amazon.smithy.dafny.conversion.ToDafny.Simple.CharacterSequence(
      s
    );
  }

  public static InternalResult<
    EncryptionContext,
    Error
  > legacyEncryptionContext(
    software.amazon.cryptography.dbencryptionsdk.dynamodb.itemencryptor.internaldafny.types.DynamoDbItemEncryptorConfig config
  ) {
    try {
      EncryptionContext.Builder encryptionContextBuilder =
        new EncryptionContext.Builder()
          .tableName(ToNativeString(config.dtor_logicalTableName()))
          .hashKeyName(ToNativeString(config.dtor_partitionKeyName()));

      final EncryptionContext encryptionContext = config
          .dtor_sortKeyName()
          .is_Some()
        ? encryptionContextBuilder
          .rangeKeyName(
            ToNativeString(config.dtor_sortKeyName().dtor_value())
          )
          .build()
        : encryptionContextBuilder.build();

      return InternalResult.success(encryptionContext);
    } catch (Exception ex) {
      return InternalResult.failure(Error.create_Opaque(ex));
    }
  }

  public static InternalResult<
    Map<String, Set<EncryptionFlags>>,
    Error
  > legacyActions(
    DafnyMap<
      ? extends DafnySequence<? extends Character>,
      ? extends CryptoAction
    > attributeActionsOnEncrypt
  ) {
    try {
      final EnumSet<EncryptionFlags> signOnly = EnumSet.of(
        EncryptionFlags.SIGN
      );
      final EnumSet<EncryptionFlags> encryptAndSign = EnumSet.of(
        EncryptionFlags.ENCRYPT,
        EncryptionFlags.SIGN
      );

      Map<String, Set<EncryptionFlags>> legacyActions = new HashMap<>();
      BiConsumer<
        ? super DafnySequence<? extends Character>,
        ? super CryptoAction
      > buildLegacyActions = (dafnyKey, value) -> {
        final String key = ToNativeString(dafnyKey);
        if (value.is_SIGN__ONLY()) {
          legacyActions.put(key, signOnly);
        } else if (value.is_ENCRYPT__AND__SIGN()) {
          legacyActions.put(key, encryptAndSign);
        } else if (value.is_DO__NOTHING()) {
          // DDB-EC v1 is not explicit
        } else {
          // Exceptional Postcondition: Only ENCRYPT and SIGN are supported. Do Nothing is not supported
          throw new IllegalArgumentException("Unsupported CryptoAction.");
        }
      };
      attributeActionsOnEncrypt.forEach(buildLegacyActions);
      return InternalResult.success(legacyActions);
    } catch (IllegalArgumentException ex) {
      final Error dafnyEx = Error.create_DynamoDbItemEncryptorException(
        ToDafnyString(ex.getMessage())
      );
      return InternalResult.failure(dafnyEx);
    } catch (Exception ex) {
      return InternalResult.failure(Error.create_Opaque(ex));
    }
  }
}

File: DynamoDbEncryption/runtimes/python/src/aws_dbesdk_dynamodb/internaldafny/extern/InternalLegacyOverride.py
# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
from _dafny import Seq

import aws_dbesdk_dynamodb.internaldafny.generated.InternalLegacyOverride
from aws_dbesdk_dynamodb.internaldafny.generated.AwsCryptographyDbEncryptionSdkDynamoDbItemEncryptorTypes import (
    DynamoDbItemEncryptorConfig_DynamoDbItemEncryptorConfig,
    Error_DynamoDbItemEncryptorException,
    Error_Opaque,
    DecryptItemInput_DecryptItemInput,
    EncryptItemInput_EncryptItemInput,
)
from aws_dbesdk_dynamodb.internaldafny.generated.AwsCryptographyDbEncryptionSdkStructuredEncryptionTypes import (
    CryptoAction_ENCRYPT__AND__SIGN,
    CryptoAction_SIGN__ONLY,
    CryptoAction_DO__NOTHING,
)
from aws_dbesdk_dynamodb.smithygenerated.aws_cryptography_dbencryptionsdk_dynamodb.references import (
    ILegacyDynamoDbEncryptor,
)
from aws_dbesdk_dynamodb.smithygenerated.aws_cryptography_dbencryptionsdk_dynamodb_itemencryptor.models import (
    EncryptItemInput,
    EncryptItemOutput,
    DecryptItemOutput,
    DecryptItemInput,
)

try:
    from dynamodb_encryption_sdk.encrypted.client import EncryptedClient
    from dynamodb_encryption_sdk.encrypted.table import EncryptedTable
    from dynamodb_encryption_sdk.encrypted.resource import EncryptedResource
    from dynamodb_encryption_sdk.encrypted.client import EncryptedPaginator
    from dynamodb_encryption_sdk.encrypted.item import encrypt_dynamodb_item, decrypt_dynamodb_item
    from dynamodb_encryption_sdk.structures import EncryptionContext, AttributeActions
    from dynamodb_encryption_sdk.identifiers import CryptoAction
    from dynamodb_encryption_sdk.encrypted import CryptoConfig
    from dynamodb_encryption_sdk.internal.identifiers import ReservedAttributes

    _HAS_LEGACY_DDBEC = True
except ImportError:
    _HAS_LEGACY_DDBEC = False


class InternalLegacyOverride(aws_dbesdk_dynamodb.internaldafny.generated.InternalLegacyOverride.InternalLegacyOverride):
    def __init__(self):
        super().__init__()
        self.crypto_config = None
        self.policy = None

    @staticmethod
    def Build(config: DynamoDbItemEncryptorConfig_DynamoDbItemEncryptorConfig):
        # Check for early return (Postcondition): If there is no legacyOverride there is nothing to do.
        if config.legacyOverride.is_None:
            return InternalLegacyOverride.CreateBuildSuccess(InternalLegacyOverride.CreateInternalLegacyOverrideNone())

        legacy_override = config.legacyOverride.value

        # Precondition: The encryptor MUST be a DynamoDBEncryptor
        if not _HAS_LEGACY_DDBEC:
            return InternalLegacyOverride.CreateBuildFailure(
                InternalLegacyOverride.CreateError("Could not find aws-dynamodb-encryption-python installation")
            )

        # Precondition: The encryptor MUST be one of the supported legacy types
        if not (
            isinstance(legacy_override.encryptor, EncryptedClient)
            or isinstance(legacy_override.encryptor, EncryptedTable)
            or isinstance(legacy_override.encryptor, EncryptedResource)
        ):
            return InternalLegacyOverride.CreateBuildFailure(
                InternalLegacyOverride.CreateError("Legacy encryptor is not supported")
            )

        # Preconditions: MUST be able to create valid encryption context
        maybe_encryption_context = InternalLegacyOverride.legacyEncryptionContext(config)
        if maybe_encryption_context.is_Failure:
            return maybe_encryption_context

        # Precondition: All actions MUST be supported types
        maybe_actions = InternalLegacyOverride.legacyActions(legacy_override.attributeActionsOnEncrypt)
        if maybe_actions.is_Failure:
            return maybe_actions

        # Create and return the legacy override instance
        legacy_instance = InternalLegacyOverride()
        legacy_instance.policy = legacy_override.policy
        legacy_instance.crypto_config = CryptoConfig(
            materials_provider=legacy_override.encryptor._materials_provider,
            encryption_context=maybe_encryption_context.value,
            attribute_actions=maybe_actions.value,
        )
        return InternalLegacyOverride.CreateBuildSuccess(
            InternalLegacyOverride.CreateInternalLegacyOverrideSome(legacy_instance)
        )

    @staticmethod
    def legacyEncryptionContext(config: DynamoDbItemEncryptorConfig_DynamoDbItemEncryptorConfig):
        """Create the legacy encryption context from the config."""
        try:
            # Convert Dafny types to Python strings for the encryption context
            table_name = InternalLegacyOverride.DafnyStringToNativeString(config.logicalTableName)
            partition_key_name = InternalLegacyOverride.DafnyStringToNativeString(config.partitionKeyName)
            sort_key_name = (
                InternalLegacyOverride.DafnyStringToNativeString(config.sortKeyName.value)
                if config.sortKeyName.is_Some
                else None
            )

            # Create the legacy encryption context with the extracted values
            encryption_context = EncryptionContext(
                table_name=table_name,
                partition_key_name=partition_key_name,
                sort_key_name=sort_key_name,
            )

            return InternalLegacyOverride.CreateBuildSuccess(encryption_context)
        except Exception as ex:
            return InternalLegacyOverride.CreateBuildFailure(Error_Opaque(ex))

    @staticmethod
    def legacyActions(attribute_actions_on_encrypt):
        """Create the legacy attribute actions from the config."""
        try:
            # Create a new AttributeActions with default ENCRYPT_AND_SIGN
            # Default Action to take if no specific action is defined in ``attribute_actions``
            # https://docs.aws.amazon.com/database-encryption-sdk/latest/devguide/DDBEC-legacy-concepts.html#legacy-attribute-actions
            legacy_actions = AttributeActions(default_action=CryptoAction.ENCRYPT_AND_SIGN)

            # Map the action from the config to legacy actions
            attribute_actions = {}
            for key, action in attribute_actions_on_encrypt.items:
                key_str = InternalLegacyOverride.DafnyStringToNativeString(key)

                # Map the action type to the appropriate CryptoAction
                if action == CryptoAction_ENCRYPT__AND__SIGN():
                    attribute_actions[key_str] = CryptoAction.ENCRYPT_AND_SIGN
                elif action == CryptoAction_SIGN__ONLY():
                    attribute_actions[key_str] = CryptoAction.SIGN_ONLY
                elif action == CryptoAction_DO__NOTHING():
                    attribute_actions[key_str] = CryptoAction.DO_NOTHING
                else:
                    return InternalLegacyOverride.CreateBuildFailure(
                        InternalLegacyOverride.CreateError(f"Unknown action type: {action}")
                    )

            # Update the attribute_actions dictionary
            legacy_actions.attribute_actions = attribute_actions
            return InternalLegacyOverride.CreateBuildSuccess(legacy_actions)
        except Exception as ex:
            return InternalLegacyOverride.CreateBuildFailure(Error_Opaque(ex))

    def EncryptItem(self, input: EncryptItemInput_EncryptItemInput):
        """Encrypt an item using the legacy DynamoDB encryptor.

        :param input: EncryptItemInput containing the plaintext item to encrypt
        :returns Result containing the encrypted item or an error
        """
        try:
            # Precondition: Policy MUST allow the caller to encrypt.
            if not self.policy.is_FORCE__LEGACY__ENCRYPT__ALLOW__LEGACY__DECRYPT:
                return self.CreateEncryptItemFailure(
                    InternalLegacyOverride.CreateError("Legacy policy does not support encrypt")
                )

            # Get the Native Plaintext Item
            native_input = aws_dbesdk_dynamodb.smithygenerated.aws_cryptography_dbencryptionsdk_dynamodb_itemencryptor.dafny_to_smithy.aws_cryptography_dbencryptionsdk_dynamodb_itemencryptor_EncryptItemInput(
                input
            )

            # Encrypt the item using the instance attributes
            encrypted_item = encrypt_dynamodb_item(
                item=native_input.plaintext_item,
                crypto_config=self.crypto_config.with_item(native_input.plaintext_item),
            )

            # Return the encrypted item
            # The legacy encryption method returns items in the format that Dafny expects,
            # so no additional conversion is needed here
            native_output = EncryptItemOutput(encrypted_item=encrypted_item, parsed_header=None)
            dafny_output = aws_dbesdk_dynamodb.smithygenerated.aws_cryptography_dbencryptionsdk_dynamodb_itemencryptor.smithy_to_dafny.aws_cryptography_dbencryptionsdk_dynamodb_itemencryptor_EncryptItemOutput(
                native_output
            )
            return self.CreateEncryptItemSuccess(dafny_output)

        except Exception as ex:
            return self.CreateEncryptItemFailure(InternalLegacyOverride.CreateError(Error_Opaque(ex)))

    def DecryptItem(self, input: DecryptItemInput_DecryptItemInput):
        """Decrypt an item using the legacy DynamoDB encryptor.

        :param input: DecryptItemInput containing the encrypted item to decrypt
        :returns Result containing the decrypted item or an error
        """
        try:
            # Precondition: Policy MUST allow the caller to decrypt.
            # = specification/dynamodb-encryption-client/decrypt-item.md#behavior
            ## If a [Legacy Policy](./ddb-table-encryption-config.md#legacy-policy) of
            ## `FORBID_LEGACY_ENCRYPT_FORBID_LEGACY_DECRYPT` is configured,
            ## and the input item [is an item written in the legacy format](#determining-legacy-items),
            ## this operation MUST fail.
            if not (
                self.policy.is_FORCE__LEGACY__ENCRYPT__ALLOW__LEGACY__DECRYPT
                or self.policy.is_FORBID__LEGACY__ENCRYPT__ALLOW__LEGACY__DECRYPT
            ):
                return self.CreateDecryptItemFailure(
                    InternalLegacyOverride.CreateError("Legacy policy does not support decrypt")
                )

            # Get the Native DecryptItemInput
            native_input: DecryptItemInput = (
                aws_dbesdk_dynamodb.smithygenerated.aws_cryptography_dbencryptionsdk_dynamodb_itemencryptor.dafny_to_smithy.aws_cryptography_dbencryptionsdk_dynamodb_itemencryptor_DecryptItemInput(
                    input
                )
            )
            # Decrypt the item using the instance attributes
            decrypted_item = decrypt_dynamodb_item(
                item=native_input.encrypted_item,
                crypto_config=self.crypto_config.with_item(native_input.encrypted_item),
            )

            native_output = DecryptItemOutput(plaintext_item=decrypted_item, parsed_header=None)
            dafny_output = aws_dbesdk_dynamodb.smithygenerated.aws_cryptography_dbencryptionsdk_dynamodb_itemencryptor.smithy_to_dafny.aws_cryptography_dbencryptionsdk_dynamodb_itemencryptor_DecryptItemOutput(
                native_output
            )
            return self.CreateDecryptItemSuccess(dafny_output)
        except Exception as ex:
            return self.CreateDecryptItemFailure(InternalLegacyOverride.CreateError(Error_Opaque(ex)))

    def IsLegacyInput(self, input: DecryptItemInput_DecryptItemInput):
        """
        Determine if the input is from a legacy client.

        :param input: The decrypt item input to check
        :returns Boolean indicating if the input is from a legacy client
        """
        if not input.is_DecryptItemInput:
            return False

        # Get the Native DecryptItemInput
        native_input: DecryptItemInput = (
            aws_dbesdk_dynamodb.smithygenerated.aws_cryptography_dbencryptionsdk_dynamodb_itemencryptor.dafny_to_smithy.aws_cryptography_dbencryptionsdk_dynamodb_itemencryptor_DecryptItemInput(
                input
            )
        )
        # = specification/dynamodb-encryption-client/decrypt-item.md#determining-legacy-items
        ## An item MUST be determined to be encrypted under the legacy format if it contains
        ## attributes for the material description and the signature.
        return (
            "*amzn-ddb-map-desc*" in native_input.encrypted_item and "*amzn-ddb-map-sig*" in native_input.encrypted_item
        )

    @staticmethod
    def DafnyStringToNativeString(dafny_input):
        return b"".join(ord(c).to_bytes(2, "big") for c in dafny_input).decode("utf-16-be")

    @staticmethod
    def NativeStringToDafnyString(native_input):
        return Seq(
            "".join([chr(int.from_bytes(pair, "big")) for pair in zip(*[iter(native_input.encode("utf-16-be"))] * 2)])
        )

    @staticmethod
    def CreateError(message):
        """Create an Error with the given message."""
        return Error_DynamoDbItemEncryptorException(InternalLegacyOverride.NativeStringToDafnyString(message))


aws_dbesdk_dynamodb.internaldafny.generated.InternalLegacyOverride.InternalLegacyOverride = InternalLegacyOverride
